forge/test/mlir/llama/test_llama_inference.py

```diff
@@ -120,3 +121,346 @@ def test_llama_inference_cache_cpu():
+
+        # CPU Inference
+        framework_output = framework_model(
+            input_ids=input_ids, attention_mask=attention_mask
+        )
+
+        # Run on TT device
+        tt_output = compiled_model(*model_inputs)
+        tt_output = [tt_out.to("cpu") for tt_out in tt_output]
+
+        # Validate TT result with Framework
+        assert all(
+            [
+                compare_with_golden_pcc(golden=fw_out, calculated=tt_out, pcc=0.99)
+                for fw_out, tt_out in zip(framework_output, tt_output)
+            ]
+        )
+
+        next_token_index = non_padding_seq_len + idx
+        next_token_logits = tt_output[0][:, next_token_index - 1, :]
+        next_token = torch.argmax(next_token_logits, dim=-1)
+
+        if next_token == tokenizer.eos_token_id:
+            break
+
+        input_ids[:, next_token_index] = next_token
+        attention_mask[:, next_token_index] = 1
+
+    # Generated text
+    generated_text = tokenizer.decode(input_ids[0], skip_special_tokens=True)
+    print("generated_text=", generated_text)
+
+
+@pytest.mark.xfail()
+def test_llama_prefill_on_cpu_decode_on_tt_cache():
+
+    # Load Llama 3B model and tokenizer
+    model_path = "openlm-research/open_llama_3b"
+    framework_model, tokenizer = load_model_and_tokenizer(
+        model_path=model_path, use_cache=True, return_dict=True
+    )
+
+    class LlamaModelWrapper(torch.nn.Module):
+        """
+        In LlamaModelWrapper class, forward function takes single input token (i.e last predicted token)
+        and past key values from the previous iteration and return logits and past key values
+        Args:
+            input_id (`torch.Tensor`) - shape of (batch_size, 1)
+            past_key_values (`List[List[torch.Tensor, torch.Tensor]]`) - key/values shape - (batch_size, num_of_key_values_heads, key_value_seq_len, head_dim)
+        Returns:
+            outputs - Logits of shape (batch_size, 1, vocab_size) and past_key_values (`List[List[torch.Tensor, torch.Tensor]]`)
+            past key/values tensor shape - (batch_size, num_of_key_values_heads, key_value_seq_len + 1, head_dim)
+        """
+
+        def __init__(self, model):
+            super().__init__()
+            self.model = model
+
+        def forward(self, input_id, past_key_values):
+            model_outputs = self.model(
+                input_ids=input_id, past_key_values=past_key_values
+            )
+            return model_outputs.logits, model_outputs.past_key_values
+
+    llama_model = LlamaModelWrapper(framework_model)
+
+    # Prepare input sentence
+    prompt = "Q: What is the largest animal?\nA:"
+    inputs = tokenizer(prompt, return_tensors="pt")
+
+    # Run Prefill on CPU with cache to get the initial logits and past key-values
+    prefill_output = framework_model(
+        input_ids=inputs.input_ids, attention_mask=inputs.attention_mask
+    )
+    next_token_logits = prefill_output.logits[:, -1, :]
```

Comment:
@chandrasekaranpradeep @nvukobratTT I think here is the problem. We take next token logits as last embeding of prefil_output. Actually this will be embeding of last padded token... We should take logit on index: seq_size - 1. Do you agree?

Reply:
In prefill stage, we are only passing the input_ids and attention_mask which doesn't include any pad_token because we are not setting any max_seq_len in tokenizer.